runtime/src/message_processor.rs

```diff
@@ -579,87 +606,128 @@ impl MessageProcessor {
-                        Some((programdata_address, account))
+            let program_account =
+                invoke_context
+                    .get_account(&callee_program_id)
+                    .ok_or_else(|| {
+                        ic_msg!(invoke_context, "Unknown program {:?}", callee_program_id);
+                        InstructionError::MissingAccount
+                    })?;
+            if !program_account.borrow().executable {
+                return Err(InstructionError::AccountNotExecutable);
+            }
+            let programdata_executable =
+                if program_account.borrow().owner == bpf_loader_upgradeable::id() {
+                    if let UpgradeableLoaderState::Program {
+                        programdata_address,
+                    } = program_account.borrow().state()?
+                    {
+                        if let Some(account) = invoke_context.get_account(&programdata_address) {
+                            Some((programdata_address, account))
+                        } else {
+                            ic_msg!(
+                                invoke_context,
+                                "Unknown upgradeable programdata account {:?}",
+                                programdata_address,
+                            );
+                            return Err(InstructionError::MissingAccount);
+                        }
                     } else {
+                        ic_msg!(
+                            invoke_context,
+                            "Upgradeable program account state not valid {:?}",
+                            callee_program_id,
+                        );
                         return Err(InstructionError::MissingAccount);
                     }
                 } else {
-                    return Err(InstructionError::MissingAccount);
-                }
-            } else {
-                None
-            };
-        let mut executable_accounts = vec![(callee_program_id, program_account)];
-        if let Some(programdata) = programdata_executable {
-            executable_accounts.push(programdata);
-        }
+                    None
+                };
+            let mut executables = vec![(callee_program_id, program_account)];
+            if let Some(programdata) = programdata_executable {
+                executables.push(programdata);
+            }
+            (
+                message,
+                executables,
+                accounts,
+                account_refs,
+                caller_privileges,
+            )
+        };
 
+        #[allow(clippy::deref_addrof)]
         MessageProcessor::process_cross_program_instruction(
             &message,
-            &executable_accounts,
+            &executables,
             &accounts,
             &caller_privileges,
-            invoke_context,
+            *(&mut *(invoke_context.borrow_mut())),
```

Comment:
This is avoidable with the following patch:
```
diff --git a/programs/bpf_loader/src/syscalls.rs b/programs/bpf_loader/src/syscalls.rs
index d5e6755e4..9c9938f44 100644
--- a/programs/bpf_loader/src/syscalls.rs
+++ b/programs/bpf_loader/src/syscalls.rs
@@ -1502,7 +1502,7 @@ fn check_authorized_program(
 fn get_upgradeable_executable(
     callee_program_id: &Pubkey,
     program_account: &RefCell<Account>,
-    invoke_context: &Ref<&mut dyn InvokeContext>,
+    invoke_context: &mut dyn InvokeContext,
 ) -> Result<Option<(Pubkey, RefCell<Account>)>, EbpfError<BPFError>> {
     if program_account.borrow().owner == bpf_loader_upgradeable::id() {
         match program_account.borrow().state() {
@@ -1552,7 +1552,7 @@ fn call<'a>(
         caller_privileges,
         abort_on_all_cpi_failures,
     ) = {
-        let invoke_context = syscall.get_context()?;
+        let mut invoke_context = syscall.get_context_mut()?;
 
         invoke_context
             .get_compute_meter()
@@ -1580,7 +1580,7 @@ fn call<'a>(
                 &instruction,
                 &keyed_account_refs,
                 &signers,
-                &invoke_context,
+                *invoke_context,
             )
             .map_err(SyscallError::InstructionError)?;
         let caller_privileges = message
@@ -1616,7 +1616,7 @@ fn call<'a>(
         })?)
         .clone();
         let programdata_executable =
-            get_upgradeable_executable(&callee_program_id, &program_account, &invoke_context)?;
+            get_upgradeable_executable(&callee_program_id, &program_account, *invoke_context)?;
         let mut executables = vec![(callee_program_id, program_account)];
         if let Some(executable) = programdata_executable {
             executables.push(executable);
diff --git a/runtime/src/message_processor.rs b/runtime/src/message_processor.rs
index d7504fccb..3c2befc77 100644
--- a/runtime/src/message_processor.rs
+++ b/runtime/src/message_processor.rs
@@ -24,7 +24,7 @@ use solana_sdk::{
     transaction::TransactionError,
 };
 use std::{
-    cell::{Ref, RefCell},
+    cell::RefCell,
     collections::HashMap,
     rc::Rc,
     sync::Arc,
@@ -531,7 +531,7 @@ impl MessageProcessor {
         instruction: &Instruction,
         keyed_accounts: &[&KeyedAccount],
         signers: &[Pubkey],
-        invoke_context: &Ref<&mut dyn InvokeContext>,
+        invoke_context: &mut dyn InvokeContext,
     ) -> Result<(Message, Pubkey, usize), InstructionError> {
         // Check for privilege escalation
         for account in instruction.accounts.iter() {
@@ -606,11 +606,7 @@ impl MessageProcessor {
         keyed_accounts: &[&KeyedAccount],
         signers_seeds: &[&[&[u8]]],
     ) -> Result<(), InstructionError> {
-        let invoke_context = RefCell::new(invoke_context);
-
         let (message, executables, accounts, account_refs, caller_privileges) = {
-            let invoke_context = invoke_context.borrow();
-
             let caller_program_id = invoke_context.get_caller()?;
 
             // Translate and verify caller's data
@@ -625,7 +621,7 @@ impl MessageProcessor {
                 .collect::<Vec<bool>>();
             caller_privileges.insert(0, false);
             let (message, callee_program_id, _) =
-                Self::create_message(&instruction, &keyed_accounts, &signers, &invoke_context)?;
+                Self::create_message(&instruction, &keyed_accounts, &signers, invoke_context)?;
             let mut accounts = vec![];
             let mut account_refs = vec![];
             'root: for account_key in message.account_keys.iter() {
@@ -704,13 +700,12 @@ impl MessageProcessor {
             &executables,
             &accounts,
             &caller_privileges,
-            *(&mut *(invoke_context.borrow_mut())),
+            invoke_context,
         )?;
 
         // Copy results back to caller
 
         {
-            let invoke_context = invoke_context.borrow();
             for (i, (account, account_ref)) in accounts.iter().zip(account_refs).enumerate() {
                 let account = account.borrow();
                 if message.is_writable(i) && !account.executable {
```

Reply:
Needs more work, borrow failure due to already borrowed mutably

Reply:
Ah dang, thanks for trying